src/shogun/mathematics/linalg/GPUVectorImpl.cpp

```diff
@@ -0,0 +1,103 @@
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *  3. Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived from
+ *     this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Authors: 2016 Pan Deng, Soumyajit De, Viktor Gal
+*/
+
+#include <shogun/mathematics/linalg/GPUVectorImpl.h>
+
+#ifdef HAVE_CXX11
+#ifdef HAVE_VIENNACL
+
+namespace shogun
+{
+
+template<class T>
+Vector<T>::GPUVectorImpl::GPUVectorImpl()
+{
+}
+
+template<class T>
+Vector<T>::GPUVectorImpl::GPUVectorImpl(T* data, index_t len)
+:m_GPUptr(new VCLMemoryArray()), m_len(len), m_offset(0)
+{
+	viennacl::backend::memory_create(*m_GPUptr, sizeof(T)*m_len,
+    		viennacl::context());
+
+	viennacl::backend::memory_write(*m_GPUptr, 0, m_len*sizeof(T), data);
+}
+
+template<class T>
+Vector<T>::GPUVectorImpl::GPUVectorImpl
+(const Vector<T>::GPUVectorImpl &array)
+{
+	m_GPUptr = array.m_GPUptr;
+	m_len = array.m_len;
+	m_offset = array.m_offset;
+}
+
+template<class T>
+typename Vector<T>::GPUVectorImpl::VCLVectorBase Vector<T>::GPUVectorImpl::GPUvec()
+{
+	return VCLVectorBase(*m_GPUptr, m_len, m_offset, 1);
+}
+
+template<class T>
+typename Vector<T>::GPUVectorImpl::VCLVector Vector<T>::GPUVectorImpl::vector()
+{
+	return VCLVector(Vector<T>::GPUVectorImpl::GPUvec());
+}
+
+template <class T>
+void Vector<T>::GPUVectorImpl::transferToCPU(T* data)
+{
+	viennacl::backend::memory_read(*m_GPUptr, m_offset*sizeof(T), m_len*sizeof(T),
+		data);
+}
+
+template <class T>
+viennacl::const_entry_proxy<T> Vector<T>::GPUVectorImpl::operator[](index_t index) const
+{
+	return viennacl::const_entry_proxy<T>(m_offset+index, *m_GPUptr);
+}
+
+template <class T>
+viennacl::entry_proxy<T> Vector<T>::GPUVectorImpl::operator[](index_t index)
+{
+	return viennacl::entry_proxy<T>(m_offset+index, *m_GPUptr);
+}
+
+template class Vector<int32_t>;
```

Comment:
what about 64 bit?
Some backends dont support it I guess? Does yours?
IF the backend supports, it should be offered here as well